scripts/functions-deploy-tests/tests.ts

```diff
@@ -0,0 +1,328 @@
+  for (const ep of output.result as Endpoint[]) {
+    const id = ep.id.replace(`${stripId}-`, "");
+    if (ep.id === id) {
+      // This endpoint is from another run. Ignore.
+      continue;
+    }
+    eps[id] = ep;
+  }
+  return eps;
+}
+
+describe("firebase deploy", function (this) {
+  this.timeout(1000_000);
+
+  const RUN_ID = genRandomId();
+
+  before(async () => {
+    expect(FIREBASE_PROJECT).to.not.be.empty;
+
+    // write up index.js to import trigger definition using unique group identifier.
+    // All exported functions will have name {hash}-{trigger} e.g. 'abcdefg-v1storage'.
+    await fs.writeFile(
+      path.join(FUNCTIONS_DIR, "index.js"),
+      `export * as ${RUN_ID} from "./fns.js";`
+    );
+  });
+
+  after(async () => {
+    await fs.unlink(path.join(FUNCTIONS_DIR, "index.js"));
+  });
+
+  it("deploys functions with runtime options", async () => {
+    const opts: Opts = {
+      v1Opts: {
+        memory: "128MB",
+        maxInstances: 42,
+        timeoutSeconds: 42,
+      },
+      v2Opts: {
+        memory: "128MiB",
+        maxInstances: 42,
+        timeoutSeconds: 42,
+        // TODO: Re-enable once https://github.com/firebase/firebase-tools/issues/4679 is fixed.
+        // cpu: 2,
+        concurrency: 42,
+      },
+      v1TqOpts: {
+        retryConfig: {
+          maxAttempts: 42,
+          maxRetrySeconds: 42,
+          maxBackoffSeconds: 42,
+          maxDoublings: 42,
+          minBackoffSeconds: 42,
+        },
+        rateLimits: {
+          maxDispatchesPerSecond: 42,
+          maxConcurrentDispatches: 42,
+        },
+      },
+      v2TqOpts: {
+        retryConfig: {
+          maxAttempts: 42,
+          maxRetrySeconds: 42,
+          maxBackoffSeconds: 42,
+          maxDoublings: 42,
+          minBackoffSeconds: 42,
+        },
+        rateLimits: {
+          maxDispatchesPerSecond: 42,
+          maxConcurrentDispatches: 42,
+        },
+      },
+      v1IdpOpts: {
+        blockingOptions: {
+          idToken: true,
+          refreshToken: true,
+          accessToken: false,
+        },
+      },
+      v2IdpOpts: {
+        idToken: true,
+        refreshToken: true,
+        accessToken: true,
+      },
+    };
+    await setOpts(opts);
+
+    const result = await cli.exec(
+      "deploy",
+      FIREBASE_PROJECT,
+      ["--only", "functions", "--non-interactive", "--force"],
+      __dirname,
+      false
+    );
+
+    expect(result.stdout, "deploy result").to.match(/Deploy complete!/);
+
+    const endpoints = await listFns(RUN_ID);
+    expect(Object.keys(endpoints).length, "number of deployed functions").to.equal(FNS_COUNT);
+
+    for (const [id, e] of Object.entries(endpoints)) {
+      if (e.platform === "gcfv1") {
+        expect(e.availableMemoryMb, `${id}.availableMemoryMb`).to.equal(128);
+        expect(e.timeoutSeconds, `${id}.timeoutSeconds`).to.equal(42);
+        expect(e.maxInstances, `${id}maxInstances`).to.equal(42);
```

Comment:
This can also be simplified for readability:

```suggestion
        expect(e).to.include({
            availableMemoryMb: 128,
            timeoutSeconds: 42.
            maxInstances: 42,
        }, `endpoint ${id} properties`);
```

(even can include a message still)

https://www.chaijs.com/api/bdd/

Reply:
Nice!